### PR TITLE
[OSX] App Store Sandbox Bookmarks API

### PR DESCRIPTION
**Pull request by mklts**
## What does the pull request do?
Adds an API to work with App Store Sandbox Bookmarks. More info here #6537

### CODE REVIEW

**Comment by Mikolaytis** [comment]
usage of the API:
- set `MacOsPlatformOptions.AppStoreSandbox` to true in `Program.cs`
- subscribe to `Application.BookmarkAdded` event and save all data info to disk, add cache of `ISandboxBookmark` in memory,  make available in your app as `Dictionary<string FilePath, byte[] bookmarkData>`.
- if you want to restore a bookmark object after app restart - get `ISandboxBookmarkFactory` from locator and create bookmarks from bookmarkData.
- cover every access to file outside of sandbox with `using(bookmark.Open()){ ...access code... }`

**Comment by Mikolaytis** [comment]
Feature is done, but not tested. I will integrate result nuget to our app, test everything, fix issues.
[WIP] flag will be removed on successful app store release.

**Comment by Mikolaytis** [comment]
@kekekeks it's a big chance that this issue is known, but in case it's not, I want to note that `RunLeakTests` have a chance to fail randomly.

failed run  = https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=15467&view=results
completed run = https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=15468&view=results

**Comment by Takoooooo** [comment]
> @kekekeks it's a big chance that this issue is known, but in case it's not, I want to note that `RunLeakTests` have a chance to fail randomly.
> 
> failed run = https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=15467&view=results
> completed run = https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=15468&view=results

Yes, this is a known issue.If you know why this can happen we would be glad to know

**Comment by Mikolaytis** [comment]
@danwalmsley

**Comment by kekekeks** [comment]
Linux has something similar for sandboxed apps https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.FileChooser but I don't see anything resembling bookmarks. Those might be tracked by the sandbox automatically.

File: src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs [hunk 1]
```diff
@@ -68,6 +68,13 @@ public class MacOSPlatformOptions
         /// </summary>
         public bool ShowInDock { get; set; } = true;
 
+        /// <summary>
+        /// If you will distribute you app in Mac App Store - it should be with sandbox enabled.
+        /// This parameter enables <see cref="Application.SandboxBookmarkAdded"/> event.
+        /// It's called when a user selects a document in OpenFileDialog or SaveFileDialog or via <see cref="Application.UrlsOpened"/> event.
+        /// </summary>
+        public bool AppStoreSandbox { get; set; } = false;
```
Comment (maxkatz6) [inline src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs:76]: I am not really familiar with macos sandboxing, but do we need this property? Is it a problem having always enabled event for sandbox? I assume it just will never be called in non-sandboxed app.

Comment (Mikolaytis) [reply src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs:76]: before calling this event I'm creating a sandbox bookmark for selected path from NSSavePanel/NSOpenPanel, I did not test it, but I think not calling sandbox apis in not sandboxed app is a good idea.

File: src/Avalonia.Controls/Platform/ISandboxBookmark.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    /// <summary>
+    /// Mac App Store Sandbox Bookmark - an object that allows to access files outside of sandbox
+    /// </summary>
+    public interface ISandboxBookmark
+    {
+        /// <summary>
+        /// File Path of the Bookmark
+        /// </summary>
+        string Url { get; }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Platform/ISandboxBookmark.cs:25]: Type "string" but name is "Url". Should it be "Uri" type with "file://" protocol?

Comment (maxkatz6) [reply src/Avalonia.Controls/Platform/ISandboxBookmark.cs:25]: Though, nevermind, we already have the same in https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/UrlOpenedEventArgs.cs

**Comment by jmacato** [comment]
Closing due to inactivity and merge conflicts. Please ping me if you're gonna work on this again.

**Comment by maxkatz6** [comment]
To be continued in #7234

**Pull request closed without merging**